src/core/pointcloud/qgspointcloudlayerundocommand.cpp

```diff
@@ -31,20 +31,34 @@ QgsPointCloudLayerUndoCommandChangeAttribute::QgsPointCloudLayerUndoCommandChang
   , mNewValue( value )
 {
   QgsPointCloudIndex index = mLayer->index();
-  const QgsPointCloudAttributeCollection allAttributes = index.attributes();
-  QgsPointCloudRequest req;
-  req.setAttributes( allAttributes );
-  // we want to iterate all points so we have the correct point indexes within the node
-  req.setIgnoreIndexFilterEnabled( true );
-  std::unique_ptr<QgsPointCloudBlock> block = index.nodeData( n, req );
-  const char *ptr = block->data();
-  block->attributes().find( attribute.name(), mAttributeOffset );
-  const int size = block->pointRecordSize();
-  for ( const int point : points )
+  QgsPointCloudEditingIndex *editIndex = static_cast<QgsPointCloudEditingIndex *>( index.get() );
+
+  if ( editIndex->mEditedNodeData.contains( n ) )
+  {
+    const QgsPointCloudAttributeCollection allAttributes = index.attributes();
+    QgsPointCloudRequest req;
+    req.setAttributes( allAttributes );
+    // we want to iterate all points so we have the correct point indexes within the node
+    req.setIgnoreIndexFilterEnabled( true );
+    std::unique_ptr<QgsPointCloudBlock> block = index.nodeData( n, req );
+    const char *ptr = block->data();
+    block->attributes().find( attribute.name(), mAttributeOffset );
+    const int size = block->pointRecordSize();
+    for ( const int point : points )
+    {
+      const int offset = point * size + mAttributeOffset;
+      const double oldValue = attribute.convertValueToDouble( ptr + offset );
+      mPointValues[point] = oldValue;
+    }
+  }
+  else
   {
-    const int offset = point * size + mAttributeOffset;
-    const double oldValue = attribute.convertValueToDouble( ptr + offset );
-    mPointValues[point] = oldValue;
+    // If this is the first time this node is edited, we don't need the previous values, we will just discard the node from the edit index when undoing
+    mFirstEditForNode = true;
+    for ( const int point : points )
+    {
+      mPointValues[point] = std::numeric_limits<double>::quiet_NaN();
+    }
```

Comment:
do we even need to insert values to mPointValues in this case? It seems unnecessary...

Reply:
We need the _keys_ as the point indexes that will be modified in the redo step, hence the bogus _values_.

I was considering using two distinct vectors, one for indexes - one for new values, but there are two drawbacks: The two vectors need to be in sync or bad things will happen, and also we later need to check if `pointIndexes.contains( i )`, so we benefit from the QHash fast lookup.